### ISSUE TITLE
Bridge the React API

### ISSUE THREAD

**Issue by treshugart** [skatejs/skatejs#916]
*This issue is written as if #872 is closed.*

### Lifecycle

It'd be nice at some level (separate package / repo?) to support the React API. There's some things that obviously don't map well. For example the following lifecycles only exist in React:

- `componentWillMount()`
- `componentWillReceiveProps()`
- `shouldComponentUpdate()`
- `componentWillUpdate()`
- `componentDidUpdate()`

The `propsChangedCallback()` callback in Skate actually has a single responsibility currently to return `true` if the component should update. To keep things simple rather than introducing other lifecycles, it can also be used to execute things when a component is updated. Therefore, it *can* be used to cover:

- `componentWillReceiveProps()`
- `shouldComponentUpdate()`
- `componentWillUpdate()`

The following lifecycles have direct counterparts:

- `constructor()` -> `constructor()`
- `render()` -> `renderCallback()`
- `componentDidMount()` -> `connectedCallback()`
- `componentDidUpdate()` -> `renderedCallback()`

Given the above, the following lifecycles won't be supported with their current semantics. They will be executed but at slightly different timings:

- `componentWillMount()` - will be executed during the `connectedCallback` *before* `componentDidMount()`
- `componentWillUnmount()` - will be executed during the `disconnectedCallback`

### PropTypes

Prop types have slightly different semantics. For example, in Skate, we cater to an HTML API and allow attribute linking, coercion and (de)serialising. In React, they exist for validation purposes. We can write a Babel plugin to transform Flow / TS types into `props`. Since React is moving away from `PropTypes`, it makes sense for this to be the only deviation, but if there is a Babel plugin, the APIs here can be synonymous.

### Other APIs

*TODO: update this section to React 16.*

#### `setState`

This can be added to the prototype and simply set props.

#### `forceUpdate`

This will have to invoke the renderer directly.

#### defaultProps

This and `propTypes` will be used to compose together the Skate prop definitions.

#### displayName

This is up to the user. The element still has to be registered using a name and they can choose to use this or not.

#### propTypes

This and `defaultProps` will be used to compose together the Skate prop definitions.

#### props

This will be exposed as a getter to return `props(elem)`.

#### state

This can return exactly what `props` returns as props and state are generally distinguished as public API vs symbols in Skate, but it's not an enforced convention.

#### createElement

Basically just proxies the `h` export.

#### findDOMNode

Just returns the host element, which we already have.

**Comment by bradleyayers** [comment]
I'm hopeful that this could open the door to allowing the TypeScript React JSX definitions to be used seamlessly for skate (assuming the user chooses to extend from skate's React-based-API class).

**Comment by treshugart** [comment]
It might also be worth exporting a `{ React }` object that has `createElement` on it.

**Comment by Hotell** [comment]
yeah that would be great. Currently I have to do that manually:

```typescript
// jsx.ts
import { h } from 'skatejs';

export default const React = { createElement: h };
```
> https://github.com/Hotell/wc-react-angular-interop/blob/master/skatejs/src/shared/jsx.ts

```typescript
// Button.tsx
import { Component,prop,define } from 'skatejs';
import React from '../../shared/jsx';
// .... now you can JSX with Typescript
```
> https://github.com/Hotell/wc-react-angular-interop/blob/master/skatejs/src/components-mdl2/button/Button.tsx

**Comment by treshugart** [comment]
Initial prototype: http://www.webpackbin.com/V1zsulR-M.

Things that need to happen:

- More `PropTypes`
- ReactDOM APIs

Some things that don't work 1:1 right now:

- `ref` callback (called once in Skate, use weak maps etc, or don't set as a property)
- `state` and `props` are currently the same thing; can be separated out but design patterns in React seem to have evolved to not need the distinction. I.e. build dumb components that are wrapped in smart components etc.

**Comment by thepian** [comment]
```
import React from 'skatejs/react-like';
```

I think it is a nice option, but I wouldn't recommend people to use it beyond getting started. It adds another layer, which always has the potential to introduce confusion, distraction, and mess with your thinking.

More how to X in SkateJS, especially if it has a reference to how the same is done in React would provide most of the same value.

**Comment by treshugart** [comment]
> More how to X in SkateJS, especially if it has a reference to how the same is done in React would provide most of the same value.

It's less of a cognitive bridge than it is a refactor bridge. I'm more interested in providing an easy path from React to Skate.

**Comment by treshugart** [comment]
Update prototype: http://www.webpackbin.com/EydrgJePM

**Comment by thepian** [comment]
After working intensively with React for 6 months I still prefer custom elements by far, but I do miss some react-isms. Having a state object and a props object that you can pass along is super convenient. The state object is easy to make, so I wouldn't suggest to have that in Skate by default. With the props being directly on the element itself you cannot easily pass the properties of  one element as the props for another

```
return <Nested {...this.props} abc="blah">;
```

**Comment by fredyagomez** [comment]
@thepian @treshugart  I am using the wrapper below to reactify any new skatejs component. After I apply this, I can go in React and use the new wrapper to pass props like I normally do `<Nested {...this.props} abc="blah">;` . It works perfectly for me so far, but I would appreciate any feedback if you have one.


```

import { createElement } from 'react';
import val from '@skatejs/val';

const h = val(createElement);

export const makeReacty = tagName => (props) => {
    let attrsObj = {}, funcObj = {}, objObj = {}, boolObj={}, numberObj={}, DaddyObj = {};
    let attrsKeys = [];
    Object.keys(props).map((item)=>{
        attrsKeys.push(item);
    })
    Object.values(props).forEach((item, index)=>{
        let name;
        switch (typeof item) {
            case 'string':
                name = attrsKeys[index];
                attrsObj[name] = item;
                break;
            case 'number':
                name = attrsKeys[index];
                numberObj[name] = item;
                break;
            case 'function':
                name = attrsKeys[index];
                funcObj[name] = item;
                break;
            case 'object':
                if (attrsKeys[index] !== 'children'){
                    name = attrsKeys[index];
                    objObj[name] = item;
                } else {
                    console.log('I am an object and children');
                }
                break;
            case 'boolean':
                name = attrsKeys[index];
                boolObj[name] = item;
                break;
        }
        
    })
    DaddyObj = Object.assign({ events: funcObj, attrs: attrsObj }, objObj, boolObj, numberObj);
    return h(tagName, DaddyObj, props.children); 
}
```

**Comment by treshugart** [comment]
The `withProps` mixin / hoc automatically applies a `props` property on the component that should do exactly what's being discussed here. It's also a setter, if that's your cup of tea.

**Comment by Hotell** [comment]
API has been rewritten in 5.x to match closely React-ish nature. can we close this @treshugart  ?

**Comment by treshugart** [comment]
@Hotell Yeah, I'm not sure if there's a need to 1:1 the APIs. Closing.
